### PR TITLE
Tweaked TS generation of DO SQlite types

### PR DESCRIPTION
**Pull request by geelen**
After #2722, we now have a final, very minimal SQLite API for durable objects. So I've just tweaked the TS generation to give a few generics to make things easier to work with. These are the changes:

```diff
diff --git a/.local/types-main/2023-07-01/index.ts b/bazel-bin/types/definitions/2023-07-01/index.ts
index 45460a4d..556a55d3 100755
--- a/.local/types-main/2023-07-01/index.ts
+++ b/bazel-bin/types/definitions/2023-07-01/index.ts
@@ -2778,20 +2778,24 @@ export declare const WebSocketPair: {
   };
 };
 export interface SqlStorage {
-  exec(query: string, ...bindings: any[]): SqlStorageCursor;
+  exec<T extends Record<string, SqlStorageValue>>(
+    query: string,
+    ...bindings: any[]
+  ): SqlStorageCursor<T>;
   get databaseSize(): number;
   Cursor: typeof SqlStorageCursor;
   Statement: typeof SqlStorageStatement;
 }
 export declare abstract class SqlStorageStatement {}
-export declare abstract class SqlStorageCursor {
-  raw(): IterableIterator<((ArrayBuffer | string | number) | null)[]>;
+export type SqlStorageValue = ArrayBuffer | string | number | null;
+export declare abstract class SqlStorageCursor<
+  T extends Record<string, SqlStorageValue>,
+> {
+  raw<U extends SqlStorageValue[]>(): IterableIterator<U>;
   get columnNames(): string[];
   get rowsRead(): number;
   get rowsWritten(): number;
-  [Symbol.iterator](): IterableIterator<
-    Record<string, (ArrayBuffer | string | number) | null>
-  >;
+  [Symbol.iterator](): IterableIterator<T>;
 }
 export interface Socket {
   get readable(): ReadableStream;
```

And the relevant diff between `2023-07-01` and `experimental`:

```diff
 export declare abstract class ScheduledEvent extends ExtendableEvent {
   readonly scheduledTime: number;
   readonly cron: string;
@@ -2782,6 +2836,8 @@ export interface SqlStorage {
     query: string,
     ...bindings: any[]
   ): SqlStorageCursor<T>;
+  prepare(query: string): SqlStorageStatement;
+  ingest(query: string): SqlStorageIngestResult;
   get databaseSize(): number;
   Cursor: typeof SqlStorageCursor;
   Statement: typeof SqlStorageStatement;
@@ -2797,6 +2853,12 @@ export declare abstract class SqlStorageCursor<
   get rowsWritten(): number;
   [Symbol.iterator](): IterableIterator<T>;
 }
+export interface SqlStorageIngestResult {
+  remainder: string;
+  rowsRead: number;
+  rowsWritten: number;
+  statementCount: number;
+}
 export interface Socket {
   get readable(): ReadableStream;
   get writable(): WritableStream;
```

### CODE REVIEW

**Comment by Frederik-Baetens** [comment]
> We don't have a way of conditionally adding them in, do we?

We do, `wrangler types --x-include-runtime` generates types based on your compat flags and bindings. I'd expect that to work out of the box if you're just using an established experimental flag, and if it doesn't, we can definitely extend that mechanism to conditionally generate types.

File: src/workerd/api/sql.h [hunk 1]
```diff
@@ -33,22 +33,25 @@ class SqlStorage final: public jsg::Object, private SqliteDatabase::Regulator {
...
       JSG_METHOD(ingest);
     }
 
     JSG_READONLY_PROTOTYPE_PROPERTY(databaseSize, getDatabaseSize);
 
     JSG_NESTED_TYPE(Cursor);
     JSG_NESTED_TYPE(Statement);
+
+    JSG_TS_OVERRIDE({
+      prepare: never;
```
Comment (kentonv) [inline src/workerd/api/sql.h:51]: I don't think we have to explicitly override these to hide them... they will only show up in then first place if someone is generating types with "experimental" enabled, in which case they should actually get those declarations, right?

Comment (geelen) [reply src/workerd/api/sql.h:51]: Oh my mistake, I was looking at the wrong output directory.

**Review by kentonv — APPROVED** [review]

**Comment by kentonv** [comment]
LGTM but CI seems unhappy...

**Pull request merged**